### PR TITLE
Allow setting ulimits for containers

### LINKED ISSUE #4717: sysctl tunables
When launching a dockerised  process as root it is possible to execute some kind of init script to set sysctl values. But setting right ulimits is actually need  the -privileged flag.
When run as user (-u flag) is there a way to do subj? I did not find any in docs.
Setting of any other stuff (like /dev/shm size, https://github.com/dotcloud/docker/issues/2606) would be great.

### PR DESCRIPTION
**Pull request by cpuguy83**
Closes #4717

Add option to set ulimit settings for containers.
Currently the container inherits from the daemon's ulimit settings.

Support for this was added in libcontainer here docker/libcontainer@7ce34f58cbd4884634a0f1f82c4aaab1168ea0de

### CODE REVIEW

File: docs/sources/reference/api/docker_remote_api_v1.16.md [hunk 1]
```diff
@@ -232,6 +233,9 @@ Json Parameters:
   -   **Devices** - A list of devices to add to the container specified in the
         form
         `{ "PathOnHost": "/dev/deviceName", "PathInContainer": "/dev/deviceName", "CgroupPermissions": "mrw"}`
+  -   **Ulimits** - A list of ulimits to be set in the container, specified as
+        `<type>: { "Soft": <soft limit>, "Hard": <hard limit> }`, for example:
+        `Ulimits: { "nofile": { "Soft": 1024, "Hard", 2048 }}`
```
Comment (icecrime) [inline docs/sources/reference/api/docker_remote_api_v1.16.md]: I think you need to revert changes to v1.16 and add apply them to v1.17 now.

**Comment by icecrime** [comment]
Talked with @crosbymichael: we believe this would get merged if you submitted an implementation. Do you want to close this one and make a new PR when ready, or add new commits here?

**Comment by jeremyeder** [comment]
I plan to test this when there's code.

**Comment by cpuguy83** [comment]
Working on it.  I'll put commits here.

**Comment by cpuguy83** [comment]
Ok, it's done, tear it up!

Note, I changed the the HostConifg->Ulimtits field to be an array instead of map.

**Comment by thaJeztah** [comment]
Ulimtits!

**Comment by crosbymichael** [comment]
You are not done yet, still needs a rebase ;)

**Comment by cpuguy83** [comment]
All because of your --read-only or ;)

**Comment by cpuguy83** [comment]
Rebased and ready.

File: daemon/execdriver/native/create.go [hunk 2]
```diff
@@ -201,6 +205,18 @@ func (d *driver) setupCgroups(container *libcontainer.Config, c *execdriver.Comm
...
+func (d *driver) setupRlimits(container *libcontainer.Config, c *execdriver.Command) error {
+	if c.Resources == nil {
+		return nil
+	}
+
+	for _, rlimit := range c.Resources.Rlimits {
+		container.Rlimits = append(container.Rlimits, libcontainer.Rlimit((*rlimit)))
+	}
+
+	return nil
```
Comment (icecrime) [inline daemon/execdriver/native/create.go]: Do we really need an `error` return? It's always `nil` here, not sure about other drivers.

File: pkg/ulimit/ulimit.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	if _, exists := ulimitNameMapping[parts[0]]; !exists {
+		return nil, fmt.Errorf("invalid ulimit type: %s", parts[0])
+	}
+
+	limitVals := strings.SplitN(parts[1], ":", 2)
+	if len(limitVals) > 2 {
+		return nil, fmt.Errorf("too many limit value arguments - %s, can only have up to two, `soft[:hard]`", parts[1])
+	}
+
+	for _, l := range limitVals {
```
Comment (icecrime) [inline pkg/ulimit/ulimit.go]: Why use this loop for validation exclusively instead of keeping the results for later?

Comment (cpuguy83) [inline pkg/ulimit/ulimit.go]: I think I can get rid of this loop all together, actually and just use the values below.

**Comment by cpuguy83** [comment]
@icecrime Updated.

**Comment by cpuguy83** [comment]
One thing to note as well... I think what will happen when default ulimits are set on the daemon is when a container is then restarted (even if it was created before this PR), those ulimit settings will be applied to it.

**Comment by jeremyeder** [comment]
I don't think I'll have a chance to test this before the 1.5 devcut.  It would be great if someone else could look at it.  Does it stand a chance to get merged for 1.5?  /cc @rhatdan

**Comment by unclejack** [comment]
@cpuguy83 We should only allow the ulimit settings to be provided in API v1.17 and newer. Can you please make the change?

**Comment by crosbymichael** [comment]
@unclejack what happens when you don't prepend a version to your url stating that you want the latest?

**Comment by pires** [comment]
+1

**Comment by ekuric** [comment]
@cpuguy83 are there docker binaries with ulimit patch uploaded somewhere? Would like to try it, thx

**Comment by cpuguy83** [comment]
https://files.container42.com/docker/builds/docker-1.4.1-dev-ulimits

File: docs/sources/reference/api/docker_remote_api.md [hunk 4]
```diff
@@ -61,13 +61,20 @@ This endpoint now returns the list current execs associated with the container (
...
 **New!**
 (`ReadonlyRootfs`) can be passed in the host config to mount the container's
 root filesystem as read only.
 
+**New!**
+You can set ulimit settings to be used within the container.
+
+`POST /containers/create`
+
+**New!**
```
Comment (gdm85) [inline docs/sources/reference/api/docker_remote_api.md]: @cpuguy83 ahem..is it a duplicate?

**Comment by ekuric** [comment]
@cpuguy83 thanks for binary.However I have small question,I cloned your docker repo and built binary from it, but still miss (--ulimit=[] ) option.Wonder is https://github.com/cpuguy83/docker correct one for this binary? Thx

**Comment by cpuguy83** [comment]
@ekuric it is, but you have to checkout the right branch. See the branch name at the top of this pr

**Comment by tiborvass** [comment]
@cpuguy83 needs rebase

**Comment by cpuguy83** [comment]
rebased and also add bump to 1.18 api.

**Comment by cpuguy83** [comment]
Rebased again, added daemon test for default settings, fixed a couple of doc issues that had slightly wrong syntax.

**Comment by rhatdan** [comment]
Excellent, we have a lot of interest in this patch also.

**Comment by crosbymichael** [comment]
Moved this to code review

**Comment by crosbymichael** [comment]
LGTM

**Comment by ekuric** [comment]
I tested this patch and below things I noticed 

1. "priority" switch 
 # docker  run -it --ulimit priority=10 fedora /bin/bash  
    invalid value "priority=10" for flag --ulimit: invalid ulimit type: priority  
    See 'docker run --help'.  

2. "as" switch 
# docker run -it --ulimit as=1024 fedora /bin/bash 

when run 2. container is not starting, and I was dropped back to host shell 

3. I suppose switches "maxlogins" and "maxsyslogins" are not applicable for usage in containers ( I do not see usage for them ) - however they also do not working, but again I am not sure they are applicable in this story at all

4. in "ulimit -a" we have also '-v'  virtual memory option, however I failed to figure out how to use it in this use case, hints welcome 

5. I also noticed that only longoptions are working. I was able to start container with below options and these values were reflected inside container once started 

# docker run -it --ulimit memlock=8196 --ulimit core=8196  --ulimit data=8192 --ulimit fsize=8192 --ulimit nofile=1024:1024 --ulimit rss=4096  --ulimit cpu=4 --ulimit nproc=1000 --ulimit locks=1000 --ulimit sigpending=100 --ulimit msgqueue=1000 --ulimit nice=100 --ulimit rtprio=100 fedora /bin/bash    

I can retest again if necessary to get more info 

Other than above this all LGTM

**Comment by cpuguy83** [comment]
It might be worth disabling RLIMIT_AS since memory limits should be handled by cgroups anyway.
I have a (naive) feeling it's crashing with AS because of some issue around the re-exec that happens when starting the container.

With `as=1024` docker reports `argument list too long`, if you bump that up to something MUCH bigger (like `as=10240`) the container starts up and then gets SIGSEV... which is documented here: http://linux.die.net/man/2/setrlimit

> Also automatic stack expansion will fail (and generate a SIGSEGV that kills the process if no alternate stack has been made available via sigaltstack(2))

**Comment by cpuguy83** [comment]
FYI, rebased and disabled RLIMIT_AS support.

**Comment by ekuric** [comment]
@cpuguy83 thank you for feedback. 
Is it intention only to have longoptions and not shortoptions implemented for this PR? 

Eg. 
 docker  run -it --ulimit -c=8196 fedora  /bin/bash
invalid value "-c=8196" for flag --ulimit invalid ulimit type: -c
See 'docker run --help'.

 docker  run -it --ulimit c=8196 fedora  /bin/bash
invalid value "c=8196" for flag --ulimit: invalid ulimit type: c
See 'docker run --help'.

while 

docker run -it --ulimit core=8192 fedora /bin/bash 
works fine.

Maybe I have wrong syntax for short options above, so if that is case, please advice correct one.I tested above scenario for all "shortopt" and above apply for all of them.

Above example pass core file size to 8192 (KB) ( 8MB ) and once inside container we have 

container 

ulimit -a | grep core  
core file size          (blocks, -c) 8
If we compare same parameter on baremetal we have 

baremetal 

ulimit -c 8192 
ulimit -a | grep core
core file size          (blocks, -c) 8192

these values are same, but my point here is, maybe to avoid eventual confusion later on we should have also inside container value in KB ( and not MB ) - here 8 and 8192? 

Will priority be included in future? ( --ulimit priority=INT )?

thank you

**Comment by cpuguy83** [comment]
Correct only long opts.

"priority" is not part of the ulimit/kernel SETRLIMIT.
There is real time priority (rtprio), but not priority. See http://linux.die.net/man/2/setrlimit for the supported options.
You should be able to get what you need for priority with the existing cpu-shares functionality.

**Comment by md5** [comment]
There's also `nice`, which may be what @ekuric was looking for.

**Comment by ekuric** [comment]
@cpuguy83 all right,thanks , @md5 yes true

**Comment by jeremyeder** [comment]
@cpuguy83 what do you think about the unit conversion that's happening between the host and in-container as @ekuric mentioned?

cpu-shares requires an orchestration-system-level implementation.  So, absent of that, nice is probably what we should recommend.

**Comment by cpuguy83** [comment]
@jeremyeder I'm not sure about the unit conversion since we're just sending uints to `SETRLIMIT`

**Comment by jfrazelle** [comment]
needs a rebase and then I think we are almost there :D

**Comment by cpuguy83** [comment]
Rebased

**Comment by jfrazelle** [comment]
LGTM ping @icecrime who had previously reviewed

**Comment by icecrime** [comment]
Only gave it a partial new look (read code, no new tests), I'm still LGTM :-)

File: daemon/config.go [hunk 5]
```diff
@@ -75,6 +77,8 @@ func (config *Config) InstallFlags() {
 	opts.IPListVar(&config.Dns, []string{"#dns", "-dns"}, "DNS server to use")
 	opts.DnsSearchListVar(&config.DnsSearch, []string{"-dns-search"}, "DNS search domains to use")
 	opts.LabelListVar(&config.Labels, []string{"-label"}, "Set key=value labels to the daemon")
+	config.Ulimits = make(map[string]*ulimit.Ulimit)
+	opts.UlimitMapVar(&config.Ulimits, []string{"-default-ulimit"}, "Set default ulimits for containers")
```
Comment (icecrime) [inline daemon/config.go]: Why use the map's address here?

File: opts/ulimit.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package opts
+
+import (
+	"fmt"
+
+	"github.com/docker/docker/pkg/ulimit"
+)
+
+type UlimitOpt struct {
+	values *map[string]*ulimit.Ulimit
```
Comment (icecrime) [inline opts/ulimit.go]: Same: I'm confused by the `*map` thing. Is that required?

File: daemon/config.go (hunk 5 above)
Comment (cpuguy83) [inline daemon/config.go]: Uniqueness.

File: opts/ulimit.go (hunk 6 above)
Comment (cpuguy83) [inline opts/ulimit.go]: Ah yes, the value passed from the cli is coming in as that.

**Comment by jfrazelle** [comment]
see also https://github.com/docker/docker/issues/6560

File: opts/ulimit.go (hunk 6 above)
Comment (cpuguy83) [inline opts/ulimit.go]: And I see where I screwed this up, was pulling in the cli values as &map instead of just map.  This is updated.

**Comment by icecrime** [comment]
LGTM (sorry it needs yet another rebase... I feel guilty)

**Comment by cpuguy83** [comment]
Rebased and tests complete.

**Comment by philips** [comment]
lgtm, I couldn't find the default for 'default-ulimit'. I assume it is 1024?

**Comment by cpuguy83** [comment]
@philips No default-defaults are set, and would still inherit from the docker daemon settings.

**Comment by philips** [comment]
@cpuguy83 makes sense, I see it in the docs now, no idea how I missed that. Thanks.

**Comment by jfrazelle** [comment]
what is this waiting on?

**Comment by liquid-sky** [comment]
How can one provide RLIM_INFINITY value as per (resource.h)[http://pubs.opengroup.org/onlinepubs/7908799/xsh/sysresource.h.html] or equivalent of providing `unlimited` value when running `ulimit` for a resource?

**Comment by cpuguy83** [comment]
@liquid-sky it is unsupported. There really isn't an unlimited, it's just a really high value.

**Comment by frol** [comment]
I cannot figure out why it happens, but it seems that resources that are limited by ulimit are counted across containers. Here is an example:

If I run one container with `--ulimit 'nproc=2'` it runs fine, but I cannot start the second container with the same `--ulimit 'nproc=2'`:

Terminal `#1`:
```bash
$ docker run -it --rm --ulimit 'nproc=2' --user nobody debian bash
nobody@7005f259a827:/$
```

Terminal `#2`:
```bash
$ docker run -it --rm --ulimit 'nproc=2' --user nobody debian bash
resource temporarily unavailable
FATA[0000] Error response from daemon: Cannot start container 589299c070779487462393fcb04df05d619d2debe1b1197c41587ee53c2283b8: [8] System error: resource temporarily unavailable
```

I tested this on Ubuntu 14.04 (kernel 3.16) and Ubuntu 12.04 (kernel 3.13), Docker 1.6.0.

**Comment by cpuguy83** [comment]
After some digging, it seems `nproc` is special and is per user.

**Comment by frol** [comment]
Note: another part of discussion is going in #6479 (Fork bomb prevention) and will probably continue there.

**Comment by ashish235** [comment]
So how to set the Ulimit? Can it be done per container wise? I seem to be clueless with the discussions happening in this thread.. Can someone help please. 
I need "ulimit -c unlimited" equivalent for docker.

**Comment by ashish235** [comment]
[root@localhost ~]# docker run --ulimit core=unlimited -t ubuntu:12.04 bash -c "ulimit -c"
invalid value "core=unlimited" for flag --ulimit: strconv.ParseInt: parsing "unlimited": invalid syntax
See 'docker run --help'.

Unlimited is not a recognized keyword here. 

My docker version is the latest 1.7. 

[root@localhost ~]# docker version
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): ba1f6c3/1.6.2
OS/Arch (client): linux/amd64
INFO[0203] GET /v1.18/version
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 786b29d
OS/Arch (server): linux/amd64


Also setting big values doesn't help here. 

[root@localhost ~]# docker run --ulimit core=99999 -t ubuntu:12.04 bash -c "ulimit -c"
97

It sets only 97.

**Comment by thaJeztah** [comment]
Hi @ashish235 please keep in mind that GitHub is an issue tracker, and not a general support forum.

A number of things I noticed; first of all, although your *daemon* is 1.7.1, your *client* is still at 1.6.2, so you might want to update that as well.

There were some fixes to ulimits in docker 1.7.x, but I'm not sure if those fixes were in the *client* or *daemon*, so this may be worth checking. (https://github.com/docker/docker/issues/12515)

W.r.t. *"unlimited"*; https://github.com/docker/docker/issues/12515#issuecomment-94271873

> `unlimited` is purposefully not support since it's not a real value, it's just a constant set to a really high number.

W.r.t. *"setting big values doesn't help here"*; please note this comment: https://github.com/docker/docker/issues/12957#issuecomment-104927796

> It sets the limit value as number of bytes, while ulimit reports the number of kb. So 32768 bytes == 32 kb

Converting 99999 to kb; https://www.google.com/search?q=99999%20bytes%20in%20kibibyte I get "97", so this looks right

**Comment by AkihiroSuda** [comment]
@cpuguy83 @cyphar @giuseppe

Do you think we can add support for `RLIMIT_AS` for rootless mode?

**Pull request merged**